clusters/resource_cluster.go

```diff
@@ -363,38 +439,54 @@ func resourceClusterUpdate(ctx context.Context, d *schema.ResourceData, c *commo
 		// don't add externally added libraries, if config has no `library {}` blocks
 		return nil
 	}
-	var libraryList libraries.ClusterLibraryList
+	var libraryList libraries.LibraryList
+	libraryList.ClusterId = clusterId
 	common.DataToStructPointer(d, clusterSchema, &libraryList)
-	librariesAPI := libraries.NewLibrariesAPI(ctx, c)
-	libsClusterStatus, err := librariesAPI.ClusterStatus(clusterID)
+	libsClusterStatus, err := w.Libraries.ClusterStatusByClusterId(ctx, clusterId)
+	if err != nil {
+		return err
+	}
+	libsToInstall, libsToUninstall := libraries.GetLibrariesToInstallAndUninstall(libraryList, libsClusterStatus)
+	clusterInfo, err = clusters.GetByClusterId(ctx, clusterId)
+	err = wrapMissingClusterError(err, d.Id())
 	if err != nil {
 		return err
 	}
-	libraryList.ClusterID = clusterID
-	libsToInstall, libsToUninstall := libraryList.Diff(libsClusterStatus)
 	if len(libsToUninstall.Libraries) > 0 || len(libsToInstall.Libraries) > 0 {
 		if !clusterInfo.IsRunningOrResizing() {
-			if _, err = clusters.StartAndGetInfo(clusterID); err != nil {
+			if _, err = clusters.StartByClusterIdAndWait(ctx, clusterId); err != nil {
 				return err
 			}
 		}
 		// clusters.StartAndGetInfo() always returns a running cluster
 		// or errors out, so we just know the cluster is active.
-		err = librariesAPI.UpdateLibraries(clusterID, libsToInstall, libsToUninstall,
-			d.Timeout(schema.TimeoutUpdate))
+		// TODO: Wait needed here?
```

Comment:
This seems like a nice improvement, but I would like to decouple behavioral changes from this PR if possible. Otherwise, if we decide we actually don't want to wait on library installation, we won't have to revert all of this work.

Reply:
This won't be a behaviour change as the inner functionality is same. Uninstall -> Install -> Wait